src/com/google/javascript/jscomp/Denormalize.java

```diff
@@ -124,4 +128,48 @@ private void maybeCollapseIntoForStatements(Node n, Node parent) {
       compiler.reportCodeChange();
     }
   }
+
+  private static int getAssignOpFromOp(Node n) {
+    switch (n.getType()) {
+      case Token.BITOR:
+        return Token.ASSIGN_BITOR;
+      case Token.BITXOR:
+        return Token.ASSIGN_BITXOR;
+      case Token.BITAND:
+        return Token.ASSIGN_BITAND;
+      case Token.LSH:
+        return Token.ASSIGN_LSH;
+      case Token.RSH:
+        return Token.ASSIGN_RSH;
+      case Token.URSH:
+        return Token.ASSIGN_URSH;
+      case Token.ADD:
+        return Token.ASSIGN_ADD;
+      case Token.SUB:
+        return Token.ASSIGN_SUB;
+      case Token.MUL:
+        return Token.ASSIGN_MUL;
+      case Token.DIV:
+        return Token.ASSIGN_DIV;
+      case Token.MOD:
+        return Token.ASSIGN_MOD;
+      default:
+        return Token.ERROR;
+    }
+  }
+
+  private void maybeCollapseAssignShorthand(Node n, Node parent) {
+    if (n.isAssign() && n.getFirstChild().isName() && n.getLastChild().hasChildren()
```

Comment:
This can create an invalid AST that contains the error token. Instead of hasChildren, you need to check that the last child is an operator that can be converted to an assignment operator, eg, using a helper method like this:

    static boolean hasCorrespondingAssignmentOp(Node n) {
      switch (n.getType()) {
        case Token.BITOR:
        case Token.BITXOR:
        case Token.BITAND:
        case Token.LSH:
        case Token.RSH:
        case Token.URSH:
        case Token.ADD:
        case Token.SUB:
        case Token.MUL:
        case Token.DIV:
        case Token.MOD:
          return true;
        default:
          return false;
      }
    }